### ISSUE TITLE
plugin link query very slow

### ISSUE THREAD

**Issue by ffnoname** [baasbox/baasbox#846]
I am using the example from the plugin documentation (https://github.com/baasbox/baasbox/wiki/PluginApi):

var invoiceId = "......";
    var links=Box.Links.find({
           where:"label='invoiceItem' and out.id=?",
           params:[invoiceId]
       });
    Box.log(links);

I have about 5000 links in my database. A query like the one above takes about 1 second. Is there anything that I can do about it? can I index the label, out.id? I have the feeling I am missing something.

**Comment by giastfader** [comment]
Yes, the actual implementation is slow because it is not able to use indexes.
Try to work directly with documents:
```Javascript
  var invoiceId="....";
  //assuming that the invoices are stored into the collection "invoice"
  var items=Box.Documents.find("invoice",{
        fields:" expand(_links.out_invoiceItem.in._node)",
        where:"id = ?",
        params:[invoiceId]
   });
   Box.log(items);  
```

**Comment by ffnoname** [comment]
u rock!

**Comment by ffnoname** [comment]
Is there also a way of having search parameters on both sides of the link (out.id="...." and in._author="...")?

**Comment by giastfader** [comment]
Of course, but the syntax will become a little more tricky.
The "out" object is implicit and is the starting doc, meanwhile to access to the 'in' object you have to specify the entire 'path':
````
var links=Box.Documents.find("invoice",{
        fields:" expand(_links.out_invoiceItem.in._node)",
        where:"id = ? and _links.out_invoiceItem.in._node._author=?",
        params:[invoiceId,authorUsername]
        });
`````

**Comment by endrelovas** [comment]
Sorry to reopen it. Links are two direction, correct? At least this is what I understood. How do we query the opposite direction (form ins to outs)? Following does not seem to work. Any suggestion here?
```javascript
var items=Box.Documents.find("invoiceItem",{
        fields:" expand(_links.in_invoice.out._node)",
        where:"id = ?",
        params:[invoiceId]
   });
```

**Comment by ffnoname** [comment]
sticking to the example of the API documentation, "invoiceItem" it not the other side of the link but the label of link. the other side in the example are "products". If you like to query the "in"-side you would be doing something like:
var items=Box.Documents.find("products",{
        fields:" expand(_links.in_invoiceItem.out._node)",
        where:"id = ?",
        params:[productId]
   });

**Comment by endrelovas** [comment]
Ahhh, OK, now I understand the concept. Thank you!

**Comment by endrelovas** [comment]
I still have some missing elements. I would like to make outer join using links. Is it possible or I need to make two separate querys? Something like this I am looking for:
```
var items=Box.Documents.find("products",{
        fields:" *, expand(_links.out_invoiceItem.in._node)",
        where:"id = ?",
        params:[productId]
   });
```
Is it possible at all?

**Comment by friedhelmensch** [comment]
I need the same thing. when I try to specify a field AND an expandend link,
exactly like in the comment above I always only get the expanded link but NOT the fields of the source object (in this case "product").
basically what I would need is a fetch plan (at least as how I understand fetchplans work). is something like this possible? at the moment I solve the problem by first retrieving the "products" and then I foreach over all returned "products" and do a seperate query per "product" to get the invoices for every single product. Is there a faster way?

**Comment by vittoriomastro** [comment]
Sorry but I have this problem:
if I use

```
 result1=Box.Documents.find("Plans",{
    fields:"expand(_links.out_Plans1.in._node)",
    where: "id=?",
    params:[planId]
    });
```

it's OK. But if I add an AND parameter for the in-node, the result is empty:

```
 result1=Box.Documents.find("Plans",{
    fields:"expand(_links.out_Plans1.in._node)",
    where: "id=? and _links.out_Plans1.in._node.Language = 'EN'",
    params:[planId]
    });
```

What is the problem??
